rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+}
+
+// DatabaseDataRow represents a single data row, containing a list field values.
+message DatabaseDataRow {
+  // Values the row field values.
+  // Value is the row field values. The values aren not expected to be the
+  // direct database type but their string representation, making it possible to
+  // display them without parsing. For example, binary data can be encoded into
+  // base64.
+  repeated string Values = 1;
+}
+
+// DatabaseResultData database results data, containing fields metadata and also
+// the data itself.
+message DatabaseResultData {
+  // FieldMetadata list of fields metadata.
+  repeated DatabaseDataFieldMetadata Fields = 1;
+  // Rows list of the field values.
+  repeated DatabaseDataRow Rows = 2;
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+
+  oneof Result {
+    // Status of the query execution. It is used when the command doesn't return
+    // any data.
+    Status status = 5;
+    // Data contains the fields and data of the query execution.
+    DatabaseResultData Data = 6;
+  }
+}
+```
+
+#### Examples
+
+* Audit event generated for a query that returns data (`SELET id, name FROM events;`):
+
+```json
+{
+  ...
+  "data": {
+    "fields": [
+      {"name": "id", "type": "INTEGER"},
+      {"name": "name", "type": "STRING"}
+    ],
+    "rows": [
+      ["1", "session.start"],
+      ["2", "session.end"]
+    ]
+  }
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true,
+    "message": "3 rows affected"
+  }
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  }
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
+will be under the `options.record_session.db` role option. It will support the
+following values:
+- `on`: Enables session recording, but only queries are recorded. This will
+  mimic the current behavior.
+- `full`: Enables session recording. Queries and responses are recorded.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
+- `recording_only`: Enables recording, but events will not be emitted, they
+  will only be present on recordings. This mode will be the same format as SSH
+  sessions, where the commands and their results are only present on the
+  recording.
```

Comment:
what mode is preferred in case of mode conflict?

For `recording_only`, I assume the "start" and "end" events will still emit? Should a toggle on audit-events be part of `record_session`?

Reply:
> what mode is preferred in case of mode conflict?

Most verbose one. `recording_only > full > on > off`

> I assume the "start" and "end" events will still emit?

They will be included, like in the SSH sessions (I'll update the text to clarify that).

> Should a toggle on audit-events event be part of record_session?

I'm still unsure about this too, because it doesn't look like a recording option. But moving this around (for example, to a cluster-level config) would reduce the ability to more dynamic configurations (for example, per protocol).

Reply:
I've removed the `full` (as we're no longer recording the data returned) and `recording_only` (considered out of this RFD scope as it is not directly related to session recording) options. We're left only with `on|off`.